### PR TITLE
Generalized R2R Loss

### PR DESCRIPTION
**Pull request by bemc22**
* Edit the R2R loss (R2R) for unsupervised denoising to handle multiple noise distributions
* update docs for R2R loss
* update test for R2R loss
* Add examples for R2R loss

TODO

- [x] R2R loss for gaussian, poisson and gamma distributions
- [x] New R2R docs
- [x] tests for R2R (preliminar proposal in da97e21)
- [x] examples for R2R

other changes

- [x] include the `NoiseModel` class between bases class on physics module

### Checks to be done before submitting your PR
- [x] `python3 -m pytest tests/` runs successfully.
- [x] `black .` runs successfully.
- [x] `make html` runs successfully (in the `docs/` directory).
- [x] Updated docstrings related to the changes (as applicable).
- [x] Added an entry to the [CHANGELOG.rst](../CHANGELOG.rst).

### CODE REVIEW

File: deepinv/loss/r2r.py [hunk 1]
```diff
@@ -1,155 +1,230 @@
...
 
-    This loss can be used for unsupervised image denoising with unorganized noisy images.
+    This loss can be used for unsupervised image denoising with unorganized noisy images, where
+    the noise model :math:`y\sim p(y\vert x)` belongs to the exponential family as:
 
-    The loss is designed for the noise model:
+    .. math::
+
+        p(y\vert x) =  h(x) \exp \left( y^\top \eta(x) - \phi(x) \right).
+
```
Comment (tachella) [inline deepinv/loss/r2r.py:21]:
```suggestion
which includes the popular Gaussian, Poisson and Gamma noise distributions.
```

File: deepinv/loss/r2r.py [hunk 2]
```diff
@@ -1,155 +1,230 @@
...
-    :math:`y` is the noisy measurement,
-    :math:`z` is the additional Gaussian noise of standard deviation :math:`\sigma`,
-    and :math:`\alpha` is a scaling factor.
+    where, :math:`R` is the trainable network, :math:`A` is the forward operator,
+    :math:`y` is the noisy measurement, and :math:`\alpha` is a scaling factor.
 
 
-    This loss is statistically equivalent to the supervised loss function defined on noisy/clean image pairs
-    according to authors in https://ieeexplore.ieee.org/document/9577798
+    This loss generalizes the Recorrupted-to-Recorrupted (R2R) loss for the natural exponential family of noise models
```
Comment (tachella) [inline deepinv/loss/r2r.py:42]: It is not very clear what is the first R2R paper and what is the GR2R one. I would do something like


is 

The loss was first introduced in the `Recorrupted2Recorrupted paper <https://ieeexplore.ieee.org/document/9577798>`_ for the specific case of Gaussian noise, formalizing the `Noise2Noisier loss <https://arxiv.org/abs/1910.11908>`_ such that it is statistically equivalent to the supervised loss function defined on noisy/clean image pairs. The loss was later extended to other exponential family noise distributions in `Generalized Recorrupted2Recorrupted <https://arxiv.org/abs/2412.04648>`_, including Poisson, Gamma and Binomial noise distributions.

File: deepinv/loss/r2r.py [hunk 3]
```diff
@@ -1,155 +1,230 @@
...
+        over multiple realizations of the added noise, i.e. :math:`\hat{x} = \frac{1}{N}\sum_{i=1}^N R(y_1^{(i)})`,
         where :math:`N>1`. This can be achieved using :meth:`adapt_model`.
 
-    :param Metric, torch.nn.Module metric: metric for calculating loss, defaults to MSE.
-    :param float sigma: standard deviation of the Gaussian noise used for the perturbation.
-    :param float alpha: scaling factor of the perturbation.
-    :param int eval_n_samples: number of samples used for the Monte Carlo approximation.
+    .. deprecated:: 0.2.3
+
+        The ``sigma`` paramater is deprecated and will be removed in future versions. Use ``noise_model`` parameter instead.
```
Comment (tachella) [inline deepinv/loss/r2r.py:59]:
```suggestion
        The ``sigma`` paramater is deprecated and will be removed in future versions. Use ``noise_model=deepinv.physics.GaussianNoise(sigma=sigma)`` instead.
```

File: deepinv/loss/r2r.py [hunk 4]
```diff
@@ -1,155 +1,230 @@
...
+        self.name = "gr2r"
         self.metric = metric
-        self.sigma = sigma
         self.alpha = alpha
         self.eval_n_samples = eval_n_samples
+        self.noise_model = noise_model
+
+        if sigma is not None:
+
+            print(
```
Comment (tachella) [inline deepinv/loss/r2r.py:104]: add at the beginning from `warnings import warn`
```suggestion
            warnings.warn( ...)
```
instead of the print

File: deepinv/loss/r2r.py [hunk 5]
```diff
@@ -1,155 +1,230 @@
...
 
-        where :math:`N\geq 1` are the number of samples used for the Monte Carlo approximation.
+        where :math:`y_1^{(i)} \sim p(y_1 \vert y, \alpha)` are i.i.d samples, and :math:`N\geq 1` are the number of samples used for the Monte Carlo approximation.
         During training (i.e. when ``model.train()``), we use only one sample, i.e. :math:`N=1`
         for computational efficiency, whereas at test time, we use multiple samples for better performance.
 
         :param torch.nn.Module model: Reconstruction model.
-        :param float sigma: standard deviation of the Gaussian noise used for the perturbation.
-        :param float alpha: scaling factor of the perturbation.
+        :param NoiseModel noise_model: Noise model of the natural exponential family.
```
Comment (tachella) [inline deepinv/loss/r2r.py:142]:
```suggestion
        :param NoiseModel noise_model: Noise model of the natural exponential family. Implemented options are :class:`deepinv.physics.GaussianNoise`, :class:`deepinv.physics.PoissonNoise` and :class:`deepinv.physics.GammaNoise` 
```

File: examples/self-supervised-learning/demo_r2r_denoising.py [hunk 6]
```diff
@@ -0,0 +1,187 @@
+r"""
+Self-supervised denoising with the Generalized R2R loss.
+====================================================================================================
+
+This example shows you how to train a denoiser network in a fully self-supervised way,
+i.e., using noisy images only via the GR2R loss, which exploits knowledge about the noise distribution.
```
Comment (tachella) [inline examples/self-supervised-learning/demo_r2r_denoising.py:6]:
```suggestion
i.e., using noisy images only via the `Generalized Recorrupted2Recorrupted (GR2R) loss <https://arxiv.org/abs/2412.04648>`_, which exploits knowledge about the noise distribution.
```

File: examples/self-supervised-learning/demo_r2r_denoising.py [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+# --------------------------------------------
+# We set :class:`deepinv.loss.GeneralizedR2RLoss` as the training loss.
+#
+# .. note::
+#
+#       There are GR2R losses for various noise distributions, which can be specified by the noise model.
+# .. note::
+#
+
+epochs = 1  # choose training epochs
```
Comment (tachella) [inline examples/self-supervised-learning/demo_r2r_denoising.py:122]: is this giving reasonable results? we should probably update a checkpoint as in the other SURE example.

**Review by tachella — CHANGES_REQUESTED** [review]
thanks for the PR!

I have mostly some documentation comments that could be improved.

My other concern is regarding the example: should we add a pretrained checkpoint to show good results with a single training epoch?

File: deepinv/loss/r2r.py (hunk 1 above)
Comment (tachella) [reply deepinv/loss/r2r.py:21]: See https://en.wikipedia.org/wiki/Exponential_family for more details on the exponential family.

File: deepinv/loss/r2r.py (hunk 2 above)
Comment (bemc22) [reply deepinv/loss/r2r.py]: thanks, update in [79f1cc3]

File: deepinv/loss/r2r.py (hunk 3 above)
Comment (bemc22) [reply deepinv/loss/r2r.py]: thanks, update in [79f1cc3]

File: deepinv/loss/r2r.py (hunk 4 above)
Comment (bemc22) [reply deepinv/loss/r2r.py]: thanks, update in [79f1cc3]

File: deepinv/loss/r2r.py (hunk 5 above)
Comment (bemc22) [reply deepinv/loss/r2r.py]: thanks, update in [79f1cc3]

File: examples/self-supervised-learning/demo_r2r_denoising.py (hunk 6 above)
Comment (bemc22) [reply examples/self-supervised-learning/demo_r2r_denoising.py:6]: thanks, update in [79f1cc3]

Comment (bemc22) [reply examples/self-supervised-learning/demo_r2r_denoising.py]: thanks, updated in [9415d9b]

File: examples/self-supervised-learning/demo_r2r_denoising.py (hunk 7 above)
Comment (bemc22) [reply examples/self-supervised-learning/demo_r2r_denoising.py:124]: Okey, I have opened a PR in the [deepinv/demo](https://huggingface.co/deepinv/demo/discussions) on Hugging Face, where I uploaded the demo weights for the GR2R Poisson experiment. If accepted, lines 138-146 can be uncommented.

File: examples/self-supervised-learning/demo_r2r_denoising.py [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+#
+# We generate a dataset of noisy images corrupted by Poisson noise.
+#
+# .. note::
+#
+#       We use a subset of the whole training set to reduce the computational load of the example.
+#       We recommend to use the whole set by setting ``n_images_max=None`` to get the best results.
+
+# defined physics
+predefined_noise_models = dict(
```
Comment (tachella) [inline examples/self-supervised-learning/demo_r2r_denoising.py:67]: maybe predefine them already with noise levels?

File: examples/self-supervised-learning/demo_r2r_denoising.py [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+)
+test_dataset = datasets.MNIST(
+    root=ORIGINAL_DATA_DIR, train=False, transform=transform, download=True
+)
+
+# %%
+# Generate a dataset of noisy images
+# ----------------------------------------------------------------------------------
+#
+# We generate a dataset of noisy images corrupted by Poisson noise.
```
Comment (tachella) [inline examples/self-supervised-learning/demo_r2r_denoising.py:59]: explain that the user can change the noise distribution

File: examples/self-supervised-learning/demo_r2r_denoising.py [hunk 10]
```diff
@@ -0,0 +1,192 @@
+r"""
+Self-supervised denoising with the Generalized R2R loss.
+====================================================================================================
+
+This example shows you how to train a denoiser network in a fully self-supervised way,
+using noisy images only via the `Generalized Recorrupted2Recorrupted (GR2R) loss <https://arxiv.org/abs/2412.04648>`_, 
+which exploits knowledge about the noise distribution.
+
```
Comment (tachella) [inline examples/self-supervised-learning/demo_r2r_denoising.py:8]: maybe say here that the user can change across different noise distributions?

File: examples/self-supervised-learning/demo_r2r_denoising.py (hunk 8 above)
Comment (bemc22) [reply examples/self-supervised-learning/demo_r2r_denoising.py:68]: thanks, updated in [adc00c9]

File: examples/self-supervised-learning/demo_r2r_denoising.py (hunk 9 above)
Comment (bemc22) [reply examples/self-supervised-learning/demo_r2r_denoising.py]: thanks, updated in [adc00c9]

File: examples/self-supervised-learning/demo_r2r_denoising.py (hunk 10 above)
Comment (bemc22) [reply examples/self-supervised-learning/demo_r2r_denoising.py]: thanks, updated in [adc00c9]

File: examples/self-supervised-learning/demo_r2r_denoising.py (hunk 7 above)
Comment (tachella) [reply examples/self-supervised-learning/demo_r2r_denoising.py:124]: merged the hf PR

**Review by tachella — APPROVED** [review]
